### ISSUE TITLE
Team Dashboard Sprint

### ISSUE THREAD

**Issue by mattkrick** [ParabolInc/action#135]
![team-view-with-channels 2x](https://cloud.githubusercontent.com/assets/5514175/17308241/dde8d37e-57ed-11e6-85ed-1d4b58c4606b.png)

It
- [x] has a column for agenda items
- [x] lets you add agenda items
- [x] is subscribed to the list of team members so we'll never need a refresh (teamMembers/:teamId)
- [x] is subscribed to projects from every team member (projects/:teamId/:teamMemberId)
- [x] is subscribed to agenda items for the current team (agenda/:teamId)
- [x] Has a "sort by" placeholder that does nothing
- [x] Has a header with a static list of avatars, powered by the `teamMembers` subscription
- [x] is grey and read-only when a meeting for that team is in progress
- [x] has a pink notification bar when a team meeting is going on
- [x] has 4 columns and all projects split across them
- [x] can create a new project in any of the 4 columns
- [x] edits project description via click (enter to commit, escape to cancel)
- [x] can assign ownership of project via click on avatar/owner name
- [x] can re-assign project status via gear menu

**Comment by jordanh** [comment]
In scope, or out of scope?

It
- [ ] can create a new project in any of the 4 columns
- [ ] edits project description via click (enter to commit, escape to cancel)
- [ ] can assign ownership of project via click on gear
- [ ] re-order priority of projects via vertical DnD within lane
- [ ] assign status of project by horizontal DnD between lanes

**Comment by mattkrick** [comment]
we'll probably end up needing 1-3, but that menu is gonna take awhile.
DnD is OOS for the MVP, as that always takes longer than you'd think.

**Comment by jordanh** [comment]
Yeah, that makes sense. I think if we add:

- [ ] can re-assign project status via gear menu

Then'll we'll be all set...

**Comment by ackernaut** [comment]
I want to replace this with a more up-to-date mockup that at least has the latest card designs.

My thinking is that status and archive are handled by the right icon (not a gear per latest design) and that changing owners happens by clicking the current owner. It makes more sense with the latest design.

**Comment by ackernaut** [comment]
![outcome-card-states-by-team 2x](https://cloud.githubusercontent.com/assets/307286/17344710/d08a1284-58c8-11e6-8f7c-0f175c880511.png)
![outcome-card-states 2x](https://cloud.githubusercontent.com/assets/307286/17344708/d0872d1c-58c8-11e6-959f-ce980913514e.png)
![team-view-with-avatars 2x](https://cloud.githubusercontent.com/assets/307286/17344709/d087a026-58c8-11e6-8ea7-f9a064e354ae.png)

**Comment by ackernaut** [comment]
A few things to note about the above mock-ups:

1. The Me/Team Dashboard concepts need iteration.
2. The `OutcomeCard` design is getting pretty close to ready for MVP.
3. Cards show by team or by owner depending on context.
4. We may feel two menus will work—one for assigning/changing owners, and another for changing status or archiving.
5. Updates on `OutcomeCard` state for actions opposed to projects coming soon (some iteration was made when designing the Agenda section of the meeting UX).

**Comment by jordanh** [comment]
So I'm thinking about that big pink banner at the top that appears when there is an active meeting. It is _technically_ possible that a user could be missing more than one action meeting at a time. That would be _so_ sad and anxiety inducing. They'd have to choose, and we'd have to tell them what's up.

**Design-wise**, I was thinking there could be two states for the banner to start with: (1) there is a single active meeting on team X and (2) there are several active meetings going on.

If we were real nice fellers, what we'd do for (2) is we'd make the teams in the side nav `hot` too to let them know what "several" meant.

**Implementation-wise**, could it be driven in the following way?

TM1 = Team Member 1
TM2 = Team Member 2

1. TM1 enters the lobby and starts the meeting
1. `startMeeting` mutation creates new `Meeting` table object and updates `Team.meetingId`

For `/me` and for `/team/:id`:

1. TM2 is late for the action meeting, logs in, and hits either dashboard.
1. Container is subbed to all teams and counts `Team` records with `meetingId` !== **null**
1. We render components appropriately

**Comment by ackernaut** [comment]
I sketched out a few quick concepts based on that UX case:

![my-outcomes-active-meeting 2x](https://cloud.githubusercontent.com/assets/307286/17557228/ff20e9f6-5edb-11e6-9dc6-46a6e1f78e30.png)
![my-outcomes-multi-active-meetings 2x](https://cloud.githubusercontent.com/assets/307286/17557229/ff21806e-5edb-11e6-9283-34531714257b.png)

**Comment by mattkrick** [comment]
@jordanh 
>Container is subbed to all teams and counts Team records with meetingId !== null

This is what I wanted to avoid (subbing to multiple `Team` objects), but maybe I was wrong & we should just put a key on each `teamSub`. The chances of someone staying in the meeting lobby while they have an action meeting going on should be pretty low, so who cares if we send them phase changes & then don't do anything with them.

**Comment by mattkrick** [comment]
@ackernaut what do you think about showing the time remaining for each meeting? My thought being that I want to join the meeting that has the most time left.
Off the bat, I'm not sure I like the pink dots next to the team names, since that draws my attention there & as the dumb blonde that I am, I don't know whether to click the pretty pink bar, or the pretty pink dot. If anything, I'd say we grey it out (but still let folks click it, if they _really_ want to get in there). That way, the pink bar is getting all my attention.

**Comment by mattkrick** [comment]
A second idea for the bar that I've seen some websites do is have multiple bars. I've been on some websites that have 3 or 4. Albeit they use them like we use toast: Verify your email, Error submitting form, Reconnecting, etc.

**Comment by jordanh** [comment]
I'm not a huge fan of the multiple bars, aesthetically speaking but I'd be open to changing my mind on that.

Quick idear for the timer (for now): how about a count-up timer (folks have been meeting for X)? Might give people an idea of how deep folks are in the meeting.

**Comment by mattkrick** [comment]
well that solves the what-if-they-go-over problem. i like it.
also agree, when you get a rainbow of bars at the top, the UX kinda goes out the window.

**Comment by ackernaut** [comment]
I like these ideas! How much of this goes into MVP—or just planning ahead with data/state model?

**Comment by mattkrick** [comment]
i say the timer follows bertie. if he makes it in, so does the timer. Otherwise, we leave it for beta.

**Comment by jordanh** [comment]
On agenda items, we'll need to make certain to sort the lists:

![image](https://cloud.githubusercontent.com/assets/103000/17629141/ec555b7a-6086-11e6-9f00-e380448894ce.png)

**Comment by mattkrick** [comment]
silly question: is sorting an alpha feature? We don't handle it for projects, either.

**Comment by jordanh** [comment]
Nah, we can skip it for now. Will be related to DnD.

**Comment by ackernaut** [comment]
I think at a minimal, they should show up in the same order in the meeting.

**Comment by jordanh** [comment]
Ah, @ackernaut has a good point. Sort agenda at least by createdAt?

**Comment by ackernaut** [comment]
I think the same would apply to projects. I would want to see the same order as a participate on my laptop, etc.

**Comment by mattkrick** [comment]
true... meh, we might as well save the sortOrder in the DB then.

**Comment by jordanh** [comment]
@mattkrick so I wired up status and ownership assignment on [here](https://github.com/ParabolInc/action/commit/2187a433ad463fee9baf875ec9006c4c5f5458f1). Absent a few little mysteries related to `redux-form`, it was easy-peasy.

Question: how do we want to handle archiving items? I think we want to add task status `ARCHIVED`, right?

So [here](https://github.com/ParabolInc/action/blob/2187a433ad463fee9baf875ec9006c4c5f5458f1/src/server/graphql/models/Task/taskSchema.js#L22):

```javascript
export const TaskStatus = new GraphQLEnumType({
  name: 'TaskStatus',
  description: 'The status of the task',
  values: {
    ACTIVE: {value: ACTIVE},
    STUCK: {value: STUCK},
    DONE: {value: DONE},
    FUTURE: {value: FUTURE},
    ARCHIVED: {value: ARCHIVED}  // <-- new
  }
});

```

**Comment by mattkrick** [comment]
that's a fair question. what's the desired action when someone clicks "archived"? does it immediately go away? where's it go?

**Comment by mattkrick** [comment]
one idea is to have an `isArchived` flag, and we only subscribe to stuff where `isArchived` is false. Then when we set it to true, we still have it on the client (somewhere), but we're no longer receiving updates on it, & when they resubscribe, it doesn't come back.

Then, we've got some "Show archived items" thing that is a query & paginates every 5 or 10 items.

**Comment by jordanh** [comment]
The UX _is_ the question. @ackernaut any thoughts on what happens when we archive a task? Where can they see archived tasks?

On the DB side, we'll need to think through how subscriptions will work to ensure:

- an initial subscription does not include archived items
- when one user archives an item, that state change is propagated to all clients

I think using a field like `isArchived` makes good sense. A solution I'm not completely happy with would look like this:

- TM1 loads `/team/team123`
- TM1 creates a task called _'birthday cake bought'_ (status = `ACTIVE`, isArchived = `false`)
- TM2 loads `/team/team123`, subscription initially populated with all tasks where isArchived = `false`
- TM2 clicks Status->Archive and two discrete mutations happen:
   - _'birthday cake bought'_ (status = `ARCHIVED`, isArchived = `false`) ⟶ also sent to TM1
   - _'birthday cake bought'_ (status = `ARCHIVED`, isArchived = `true`) ⟶ not also sent to TM1

We'd need two mutations as the subscription would filter all updates where `isArchived = `true``.

What do you think? Can we do what we need to do (initial load case, change propagation case) in a single mutation somehow?

**Comment by mattkrick** [comment]
So let's start with the goal UX:

-When TM2 archives a task, it disappears & can only be seen when he clicks some link to "archived tasks", which will probably be a paginated query.
- When this mutation occurs, the same thing should happen for TM1 + a toast that says "TM2 archived task X"

So, within our server mutation, we set `isArchived = true`. That will trigger the changefeed to send a `remove` to all clients. Then, we just write some `remove` logic in our cashay subscriber telling it to pop a toast.

Now, let's say TM2 made a mistake & wants to undo that action.
- He clicks "show archived projects"
- A list of the most recent 5 archived projects is fetched & shows up.
- He clicks `unarchive`
- This sets `isArchived = false`
- We now have an `add` item coming from the changefeed.

**Comment by ackernaut** [comment]
@jordanh Will definitely get archived flow into the queue! Thanks for bring that up.

**Comment by jordanh** [comment]
Closed by #181, #183, #185, and #186.
